lib/shoryuken/message.rb

```diff
@@ -28,6 +32,8 @@ def visibility_timeout=(timeout)
         receipt_handle: data.receipt_handle,
         visibility_timeout: timeout
       )
+      @visibility_timeout = timeout
+      self.become_available_at = Time.now + timeout
     end
```

Comment:
@moyuanhuang WDYT about just having a getter for `visibility_timeout` instead of introducing a new accessor  `become_available_at`?


```ruby
def visibility_timeout
  @visibility_timeout ||= Shoryuken::Client.queues(queue).visibility_timeout
end

def visibility_timeout=(timeout)
  client.change_message_visibility(
    queue_url: queue_url,
    receipt_handle: data.receipt_handle,
    visibility_timeout: timeout
  )
  @visibility_timeout = timeout
end
```

In that way, `visibility_timeout` would also be lazy-loaded. So we wouldn't call `Shoryuken::Client.queues(queue).visibility_timeout` unless `Message#visibility_timeout` gets called.

Reply:
Sorry for the late reply! Got distracted by some other stuffs.

According to the [aws-doc], the visibility timeout, whenever set, is relevant to the time the `ChangeMessageVisibility` request is issued.

Here, the `become_available_at` stands for the **timestamp** when the message becomes available, it is computed according to

```
Time.now + timeout
```

...where `timeout` is the visibility timeout **duration** passed into this method.

This being said, I think in the `Timing` class we should either compare
- `ended_at` (a timestamp) with `sqs_msg.become_available_at` (also a timestamp), or
- `total_time / 1000.0` (a duration) with `sqs_msg.visibility_timeout - sqs_msg.visibility_timeout_reset_at` (also a duration), where `visibility_timeout_reset_at` would be an extra attribute on `Message` that we keep track of.

[aws-doc]:
https://docs.aws.amazon.com/AWSSimpleQueueService/latest/SQSDeveloperGuide/sqs-visibility-timeout.html#changing-message-visibility-timeout

Reply:
Right, what I was trying to suggest is a code change to minimize the diff in this pull request and not introducing a new accessor that is not used in SQS itself. 

The change I suggested would introduce minimum changes in `message.rb` and also in `timing.rb` as commented in [here](https://github.com/phstc/shoryuken/pull/592/files#r365613536).

Reply:
reading better what you said, I missed `Time.now + timeout`, we need to keep accumulating the total visibility timeout.

Reply:
Is there a way we can take this `@visibility_timeout = Shoryuken::Client.queues(queue).visibility_timeout` from the initializer and lazy-load it?